vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,494 @@
+	StdinPipe() (io.WriteCloser, error)
+	StdoutPipe() (io.ReadCloser, error)
+	StderrPipe() (io.ReadCloser, error)
+	Start() error
+	Wait() error
+}
+
+func makeDefaultParams() *proxyAppParams {
+	return &proxyAppParams{
+		CommandRunner:   osutilCommand,
+		MakePipe:        osutil.LongPipe,
+		InitRetryDelay:  10 * time.Second,
+		MaxInitAttempts: math.MaxInt64,
+	}
+}
+
+func init() {
+	vmimpl.Register(
+		"proxyapp",
+		func(env *vmimpl.Env) (vmimpl.Pool, error) {
+			return ctor(makeDefaultParams(), env)
+		},
+		false)
+}
+
+type pool struct {
+	mu              sync.Mutex
+	env             *vmimpl.Env
+	proxy           *ProxyApp
+	ctxRunner       context.Context
+	ctxRunnerCancel context.CancelFunc
+	count           int
+}
+
+func (pool *pool) reset() {
+	pool.cancelRun()
+	pool.ctxRunner, pool.ctxRunnerCancel = context.WithCancel(context.TODO())
+	pool.proxy = nil
+	pool.count = 0
+}
+
+func (pool *pool) cancelRun() {
+	if pool.ctxRunnerCancel != nil {
+		pool.ctxRunnerCancel()
+		pool.ctxRunnerCancel = nil
+	}
+}
+
+func initPool(params *proxyAppParams, pool *pool, command string, pluginConfig []byte, env *vmimpl.Env,
+) error {
+	pool.reset()
+
+	var err error
+	pool.proxy, err = runProxyApp(pool.ctxRunner, params, command)
+	if err != nil {
+		return fmt.Errorf("failed to run ProxyApp: %w", err)
+	}
+
+	pool.count, err = pool.proxy.CreatePool(string(pluginConfig), env.Debug)
+	if err != nil {
+		pool.cancelRun()
+		return fmt.Errorf("failed to construct pool: %w", err)
+	}
+
+	return nil
+}
+
+type Config struct {
+	Command        string          `json:"cmd"`
+	ProxyAppConfig json.RawMessage `json:"config"`
+}
+
+func parseConfig(conf []byte) (*Config, error) {
+	vmCfg := new(Config)
+	if err := config.LoadData(conf, vmCfg); err != nil {
+		return nil, fmt.Errorf("failed to parseConfig(): %w", err)
+	}
+	return vmCfg, nil
+}
+
+// Package configuration VARs are mostly needed for tests.
+type proxyAppParams struct {
+	CommandRunner   func(bin string, args ...string) subProcessCmd
+	MakePipe        func() (io.ReadCloser, io.WriteCloser, error)
+	InitRetryDelay  time.Duration
+	MaxInitAttempts int64
+}
+
+func ctor(params *proxyAppParams, env *vmimpl.Env) (vmimpl.Pool, error) {
+	subConfig, err := parseConfig(env.Config)
+	if err != nil {
+		return nil, fmt.Errorf("config parse error: %w", err)
+	}
+
+	pool := &pool{
+		env: env,
+	}
+
+	minWait := time.After(params.InitRetryDelay)
+	attempt := int64(0)
+
+	pool.mu.Lock()
+
+	go func() {
+		for {
+			attempt++
+			if attempt > params.MaxInitAttempts {
+				pool.mu.Unlock()
+				return
+			}
+
+			err = initPool(params, pool, subConfig.Command, subConfig.ProxyAppConfig, env)
+			if err != nil {
+				log.Logf(0, "failed to initPool: %v", err)
+			} else {
+				pool.mu.Unlock()
+				<-pool.proxy.onTerminated
+				pool.mu.Lock()
+			}
+			<-minWait
+			minWait = time.After(params.InitRetryDelay)
+		}
+	}()
+
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	if err != nil {
```

Comment:
> Used mutex + shared variable.
> I'd like internal goroutine to work much longer than ctor. It means I'll have no reader for the errors channel.

This is where you can use select+default clause:
```
initResult := make(chan error, 1)
go func() {
  ...
  select {
  case initResult <- err:
  default:
  }
  ...
}()
if err := <-initResult; err != nil {
  return nil, err
}
```

Reply:
Channel size.... Nice, thanks!